constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,447 @@
+        rightIndex = nextRightIndex;
+      }
+      var finalInterval = Interval.between(startTime, startInclusivity, endTime, endInclusivity);
+      if (finalInterval.isEmpty()) continue;
+
+      var newValue = transform.apply(leftValue, rightValue);
+      if (newValue.isPresent()) {
+        if (!newValue.equals(previousValue)) {
+          result.add(Segment.of(finalInterval, newValue.get()));
+        } else {
+          var previousInterval = result.remove(result.size() - 1).interval();
+          result.add(
+              Segment.of(
+                  Interval.unify(previousInterval, finalInterval),
+                  newValue.get()
+              )
+          );
+        }
+      }
+      previousValue = newValue;
+      startTime = endTime;
+      startInclusivity = endInclusivity.opposite();
+    }
+
+    return resultMap;
+  }
+
+  /**
+   * The left bound of the leftmost defined interval.
+   * @return the leftmost bound and its inclusivity
+   */
+  public Optional<Pair<Duration, Interval.Inclusivity>> minValidTimePoint() {
+    if(!this.isEmpty()) {
+      final var window = this.segments.get(0).interval();
+      return Optional.of(Pair.of(window.start, window.startInclusivity));
+    } else{
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * The right bound of the rightmost defined interval.
+   * @return the rightmost bound and its inclusivity
+   */
+  public Optional<Pair<Duration, Interval.Inclusivity>> maxValidTimePoint() {
+    if(!isEmpty()) {
+      final var window = this.segments.get(this.segments.size() - 1).interval();
+      return Optional.of(Pair.of(window.end, window.endInclusivity));
+    } else{
+      return Optional.empty();
+    }
+  }
+
+  private Interval getInterval(final int index) {
+    final var i = (index >= 0) ? index : this.segments.size() - index;
+    return this.segments.get(i).interval();
+  }
+
+  private V getValue(final int index) {
+    final var i = (index >= 0) ? index : this.segments.size() - index;
+    return this.segments.get(i).value();
+  }
+
+  public Segment<V> get(final int index) {
+    final var i = (index >= 0) ? index : this.segments.size() - index;
+    return this.segments.get(i);
+  }
+
+  public int size() {
+    return segments.size();
+  }
+
+  public boolean isEmpty() {
+    return segments.isEmpty();
+  }
+
+  @Override
+  public Iterator<Segment<V>> iterator() {
+    return this.segments.iterator();
+  }
+
+  public Iterable<Interval> iterateEqualTo(final V value) {
+    return () -> segments
+        .stream()
+        .filter($ -> $.value().equals(value))
+        .map(Segment::interval)
+        .iterator();
+  }
+
+  public Spliterator<Interval> spliterateEqualTo(final V value) {
+    return StreamSupport.stream(this.spliterator(), false).flatMap(pair -> {
+      if (pair.value().equals(value)) {
+        return Stream.of(pair.interval());
+      } else {
+        return Stream.of();
+      }
+    }).spliterator();
+  }
```

Comment:
The `Iterable` interface already provides a defaulted `spliterator` method. It seems like we could remove this method entirely in favor of `iterateEqualTo(value).spliterator()`, unless there are semantic or performance differences you want to preserve here.

Reply:
Just saw that too. Removing it now